### PR TITLE
kola/tests: add rpmostree package

### PR DESCRIPTION
**Pull request by miabbott**
Adds tests for the following `rpm-ostree` commands:
  - status
  - upgrade
  - rollback
  - cleanup
  - install
  - uninstall

The `install-uninstall` test requires network access to fetch an RPM
from EPEL.

Signed-off-by: Micah Abbott <miabbott@redhat.com>

### CODE REVIEW

**Comment by miabbott** [comment]
I opted to write the tests natively, rather than trying to reuse the `vmcheck` test suite as discussed in #911.  It was faster to do it this way, in my opinion.

Based on the feedback in #911, it seems like we should still explore reusing the upstream tests in some fashion, but this fills a gap in the meantime.

File: kola/tests/rpmostree/status.go [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+// Regex to match the likes of "rhcos:openshift/3.10/x86_64/os"
+//
+// Groups:
+// 0 - rhcos
+// 1 - openshift
+// 2 - 3.10
+// 3 - x86_64
+// 4 - os
+var rhcosOriginRegex string = `^(\w+):(\w+)/(\d\.\d+)/(\w+)/(\w+)`
```
Comment (cgwalters) [inline kola/tests/rpmostree/status.go]: Let's please not hardcode this - not just because we'd like these tests to work on FCOS too, but also because doing it here will make it *extremely* painful to change - and I'd like to change RHCOS to use a [custom origin](https://github.com/ashcrow/pivot/pull/12) for the initial state too.

If this landed as is, then in order to change RHCOS we'd have to have a patch ready for both this repo and the config, then build an assembler container containing this change, which we'd then have to carefully only use for the *new* OS version.

Basically if one has separate git repos that interlock on each other, changes become extremely hard.

File: kola/tests/rpmostree/deployments.go [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+
+	originalStatus, err := getRpmOstreeStatusJSON(c, m)
+	if err != nil {
+		c.Fatal(err)
+	}
+
+	// create a local branch to act as our upgrade target
+	originalOrigin := originalStatus.Deployments[0].Origin
+	originalCsum := originalStatus.Deployments[0].Checksum
+	createBranch := "sudo ostree refs --create local-branch " + originalCsum
```
Comment (cgwalters) [inline kola/tests/rpmostree/deployments.go]: This isn't a problem in *this* code but something to watch out for is shell quoting.  I'm not sure if kola or Go have an equivalent of [this API](https://developer.gnome.org/glib/stable/glib-Shell-related-Utilities.html#g-shell-quote).  Basically if some value can e.g. have spaces or shell metacharacters itself.

File: kola/tests/rpmostree/deployments.go [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+		c.Fatal(err)
+	}
+
+	// should have two deployments
+	if len(postUpgradeStatus.Deployments) != 2 {
+		c.Fatalf("Expected two deployments; found %d deployments", len(postUpgradeStatus.Deployments))
+	}
+
+	// reboot into new deployment
+	rebootErr := m.Reboot()
```
Comment (cgwalters) [inline kola/tests/rpmostree/deployments.go]: Hooray a reboot API!

File: kola/tests/rpmostree/deployments.go [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+	// version should be new version string
+	if postRebootStatus.Deployments[0].Version != "kola-test-1.0" {
+		c.Fatalf(`New deployment version is incorrect; expected %q, got %q`, "kola-test-1.0", postRebootStatus.Deployments[0].Checksum)
+	}
+
+	// fix original origin file to point to proper ref
+	originalOriginFile := "/ostree/deploy/rhcos/deploy/" + originalCsum + ".0.origin"
+	newSedCmd := "sudo sed -i 's|^refspec=.*$|refspec=" + originalOrigin + "|' " + originalOriginFile
+	c.MustSSH(m, newSedCmd)
```
Comment (cgwalters) [inline kola/tests/rpmostree/deployments.go]: And then this wouldn't be necessary if we rebased initially right?

File: kola/tests/rpmostree/deployments.go [hunk 5]
```diff
@@ -0,0 +1,300 @@
...
+func rpmOstreeUpgradeRollback(c cluster.TestCluster) {
+	m := c.Machines()[0]
+
+	originalStatus, err := getRpmOstreeStatusJSON(c, m)
+	if err != nil {
+		c.Fatal(err)
+	}
+
+	// create a local branch to act as our upgrade target
+	originalOrigin := originalStatus.Deployments[0].Origin
```
Comment (arithx) [inline kola/tests/rpmostree/deployments.go]: Each time you reference `Deployments[0]` (ex: here, `postRebootStatus` & `originalStatus` in `rpmOstreeInstallUninstall`) it would be safer to add a `len(<status_var>.Deployments) > 0` safety check to prevent panic's if unexpected results occur.

File: kola/tests/rpmostree/status.go (hunk 1 above)
Comment (miabbott) [reply kola/tests/rpmostree/status.go]: Got it; will remove this in an updated commit

File: kola/tests/rpmostree/deployments.go [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+	// e.g. fpaste-0.3.7.4.1-2.el7.noarch
+	rpmOut := c.MustSSH(m, "rpm -q fpaste")
+	rpmMatch := regexp.MustCompile("^fpaste.*noarch").MatchString(string(rpmOut))
+	if !rpmMatch {
+		c.Fatalf(`Output from "rpm -q" was unexpected: %q`, string(rpmOut))
+	}
+
+	// e.g. fpaste 0.3.7.4.1
+	verOut := c.MustSSH(m, "fpaste --version")
+	verMatch := regexp.MustCompile(`^fpaste \d+\.\d+\.\d+\.\d+\.\d+`).MatchString(string(verOut))
```
Comment (cgwalters) [inline kola/tests/rpmostree/deployments.go]: Why test the versioning scheme?  Doesn't seem worth it to potentially have a fire drill if fpaste changes its versioning scheme.  Maybe just test `fpaste --version` exits successfully?

File: kola/tests/rpmostree/deployments.go [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+
+	// create a local branch to act as our upgrade target
+	originalOrigin := originalStatus.Deployments[0].Origin
+	originalCsum := originalStatus.Deployments[0].Checksum
+	createBranch := "sudo ostree refs --create local-branch " + originalCsum
+	c.MustSSH(m, createBranch)
+
+	// make a commit to the new branch
+	refspec := strings.Split(originalStatus.Deployments[0].Origin, ":")[1]
+	createCommit := "sudo ostree commit -b local-branch --tree ref=" + refspec + " --add-metadata-string version=kola-test-1.0"
```
[the remark this answers is not in the document]
Comment (miabbott) [reply kola/tests/rpmostree/deployments.go]: I was emulating what we did in a-h-t for this kind of test; didn't think to use commit ID

File: kola/tests/rpmostree/status.go [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+		Run:         rpmOstreeStatus,
+		ClusterSize: 1,
+		Name:        "rhcos.rpmostree.status",
+		Distros:     []string{"rhcos"},
+	})
+}
+
+// hard code the osname for RHCOS
+// TODO: should this also support FCOS?
+var rhcosOsname string = "rhcos"
```
Comment (arithx) [inline kola/tests/rpmostree/status.go]: You could wrap up these three var's inside one block like so:

```golang
var (
  rhcosOsname = "rhcos"

  // comments here
  rhcosOriginRegex = ...
)

File: kola/tests/rpmostree/deployments.go [hunk 9]
```diff
@@ -0,0 +1,300 @@
...
+	newCommit := c.MustSSH(m, createCommit)
+
+	// edit origin file to point to new branch
+	currentDir := c.MustSSH(m, "sudo ostree admin --print-current-dir")
+	originFile := string(currentDir) + ".origin"
+	sedCmd := "sudo sed -i 's|^refspec=.*$|refspec=local-branch|' " + originFile
+	c.MustSSH(m, sedCmd)
+
+	// use rpm-ostree upgrade to get to the "new" commit
+	c.MustSSH(m, "sudo rpm-ostree upgrade")
```
[the remark this answers is not in the document]
Comment (miabbott) [reply kola/tests/rpmostree/deployments.go]: I was emulating what we did in a-h-t; I guess if the end result is the same I could use `rebase`, but I was trying to tickle the `upgrade` code path.

File: kola/tests/rpmostree/deployments.go (hunk 2 above)
Comment (arithx) [reply kola/tests/rpmostree/deployments.go]: I don't believe that there is anything in the stdlib but there are definitely external libs for it.

File: kola/tests/rpmostree/deployments.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (cgwalters) [reply kola/tests/rpmostree/deployments.go]: Understood, but internally rpm-ostree is really doing almost exactly the same thing.  `upgrade`  along with most other verbs get unrolled into the generic `UpdateDeployment` DBus API.

One thing missing with this "sed behind rpm-ostree's back" approach is that it might be racy without `rpm-ostree reload`.  I also consider the `.origin` file format mostly private to ostree - in theory we could change, though it is unlikely.

[the remark this answers is not in the document]
Comment (cgwalters) [reply kola/tests/rpmostree/deployments.go]: (Although there are differences between `upgrade` and `rebase` - one of those is that the latter doesn't check that the timestamp on the commit is newer, but we're not testing that here)

[the remark this answers is not in the document]
Comment (miabbott) [reply kola/tests/rpmostree/deployments.go]: OK, since it's basically the same, I'll go with `rebase`.  Thanks for the additional details!

**Comment by miabbott** [comment]
Updated to address comments from initial feedback! :arrow_up:

File: kola/tests/rpmostree/deployments.go [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+	}
+
+	// validate we are back to the original deployment by comparing the
+	// the two rpmOstreeDeployment structs
+	if !reflect.DeepEqual(originalStatus.Deployments[0], rollbackStatus.Deployments[0]) {
+		c.Fatalf(`Differences found in "rpm-ostree status"; original %v, current: %v`, originalStatus.Deployments[0], rollbackStatus.Deployments[0])
+	}
+
+	// cleanup our mess
+	cleanupErr := rpmOstreeCleanup(c, m)
```
Comment (arithx) [inline kola/tests/rpmostree/deployments.go]: Unless we're wanting to test the cleanup we can drop this as the machine will just be terminated after the test is run (same for this in `rpmOstreeInstallUninstall`).

Comment (miabbott) [reply kola/tests/rpmostree/deployments.go]: So if I `kola run ... rhcos.rpmostree.*`, a new machine is started for each of the tests?

Comment (arithx) [reply kola/tests/rpmostree/deployments.go]: That's correct. Only `NativeFuncs` (e.x.: `rhcos.basic`) in the same test are ran on the same machine.

Comment (miabbott) [reply kola/tests/rpmostree/deployments.go]: Huh...TIL.  :smile:

Comment (miabbott) [reply kola/tests/rpmostree/deployments.go]: I think I'll keep those in there just as a sanity test of `cleanup`

File: kola/tests/rpmostree/deployments.go [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+
+	m := c.Machines()[0]
+
+	originalStatus, err := getRpmOstreeStatusJSON(c, m)
+	if err != nil {
+		c.Fatal(err)
+	}
+
+	if len(originalStatus.Deployments) < 1 {
+		c.Fatalf(`Unexpected results from "rpm-ostree status"; no deployments present?`)
```
[the remark this answers is not in the document]
Comment (miabbott) [reply kola/tests/rpmostree/deployments.go]: Added in new commit.

**Comment by miabbott** [comment]
I pushed a new commit yesterday that had one small change suggested.  :arrow_up:

**Comment by miabbott** [comment]
good to merge
